rbs/prism/AssertionsRewriterPrism.cc

```diff
@@ -1,16 +1,1140 @@
+ *
+ * Given a case like this one:
+ *
+ *     #: [X] (X) -> void
+ *     def foo(x)
+ *       y = nil #: X?
+ *     end
+ *
+ * We need to be aware of the type parameter `X` so we can use it to resolve the type of `y`.
+ */
+vector<pair<core::LocOffsets, core::NameRef>>
+extractTypeParamsPrism(core::MutableContext ctx, const parser::Prism::Parser &parser, pm_node_t *block) {
+    vector<pair<core::LocOffsets, core::NameRef>> typeParams;
+
+    // Do we have a block?
+    if (!block || !PM_NODE_TYPE_P(block, PM_BLOCK_NODE)) {
+        return typeParams;
+    }
+
+    auto *blockNode = down_cast<pm_block_node_t>(block);
+    if (!blockNode->body) {
+        return typeParams;
+    }
+
+    // The body can be either a StatementsNode or directly a CallNode (from SigsRewriter)
+    pm_node_t *node = blockNode->body;
+
+    // If it's a statements node, get the first statement
+    if (PM_NODE_TYPE_P(node, PM_STATEMENTS_NODE)) {
+        auto *statements = down_cast<pm_statements_node_t>(node);
+        ENFORCE(statements->body.size > 0);
+        node = statements->body.nodes[0];
+    }
+
+    pm_call_node_t *call = nullptr;
+
+    // Walk through the call chain to find type_parameters()
+    while (node && PM_NODE_TYPE_P(node, PM_CALL_NODE)) {
+        auto *callNode = down_cast<pm_call_node_t>(node);
+        auto methodName = parser.resolveConstant(callNode->name);
+
+        if (methodName == "type_parameters") {
+            call = callNode;
+            break;
+        }
+
+        node = callNode->receiver;
+    }
+
+    if (call == nullptr) {
+        return typeParams;
+    }
+
+    // Collect the type parameters from the arguments
+    if (auto *args = call->arguments) {
+        for (size_t i = 0; i < args->arguments.size; i++) {
+            pm_node_t *arg = args->arguments.nodes[i];
+            if (!PM_NODE_TYPE_P(arg, PM_SYMBOL_NODE)) {
+                continue;
+            }
+
+            auto *sym = down_cast<pm_symbol_node_t>(arg);
+            auto symbolName = parser.extractString(&sym->unescaped);
+            auto nameRef = ctx.state.enterNameUTF8(symbolName);
+            auto loc = parser.translateLocation(arg->location);
+            typeParams.emplace_back(loc, nameRef);
+        }
+    }
+
+    return typeParams;
+}
+
+/**
+ * Check if two Prism nodes refer to the same constant.
+ *
+ * Recursively compares constant names and their scopes.
+ * Handles both PM_CONSTANT_READ_NODE (e.g., `G2`) and PM_CONSTANT_PATH_NODE (e.g., `T::Array`).
+ */
+bool sameConstant(parser::Prism::Parser &parser, pm_node_t *a, pm_node_t *b) {
+    if (a == nullptr || b == nullptr) {
+        return false;
+    }
+
+    pm_node_type aType = PM_NODE_TYPE(a);
+    pm_node_type bType = PM_NODE_TYPE(b);
+
+    if (!((aType == PM_CONSTANT_READ_NODE || aType == PM_CONSTANT_PATH_NODE) &&
+          (bType == PM_CONSTANT_READ_NODE || bType == PM_CONSTANT_PATH_NODE))) {
+        return false;
+    }
```

Comment:
This looks to be redundant to the two checks below. They'll both be false, and you just fall off the end and `return false`